Use test fixtures (continued), organize tests by function

w-bonelli opened the pull request:
This work-in-progress PR builds on [1450](https://github.com/modflowpy/flopy/pull/1450).

~~I can stop here before moving on with more test cases. Alternatively I can continue working through tests and add on to this PR.~~ Following discussion with @langevin-usgs I will keep updating this PR as a WIP. Will request review when ready for merge. Any feedback on approach or naming is welcome and appreciated. It was also agreed that tests need not be marked by the name of the script they used to live in (e.g., `@pytest.mark.t007`). If it is preferable to keep the previous association explicit for some subset of the tests please let me know, otherwise I'll continue to omit the previous location markers.

langevin-usgs commented on autotest/test_igr.py:
```diff
@@ -11,24 +11,23 @@
 import pytest
```
Just an FYI that this file should be test_lgr.py (instead of test_igr.py), which stands for Local Grid Refinement.

w-bonelli replied:
Got it- fixed the name

mwtoews commented on .github/workflows/ci.yml:
```diff
@@ -162,7 +162,7 @@ jobs:
       working-directory: ./autotest
       run: |
-        pytest -v ci_prepare.py
+        get-modflow "$HOME/.local/bin"
```
Use something like this:
https://github.com/modflowpy/flopy/blob/3e2b5fb0ec0fc4b58854fce6155eaa40c25603c6/.github/workflows/rtd.yml#L54-L56

w-bonelli replied:
Thanks!

mwtoews commented on .github/workflows/ci-windows.yml:
```diff
@@ -100,10 +100,12 @@ jobs:
       working-directory: ./autotest
       run: |
-        pytest -v ci_prepare.py
+        get-modflow "C:\Users\runneradmin\.local\bin" --token $env:GITHUB_TOKEN
```
Use [md](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/md) to ensure directory exists:
```
md C:\Users\runneradmin\.local\bin
```

w-bonelli replied:
added

mwtoews commented on flopy/utils/get_modflow.py:
```diff
@@ -179,8 +181,15 @@ def run_main(
         req_url = f"{api_url}/releases/tags/{release_id}"
     try:
-        with urllib.request.urlopen(req_url) as resp:
+        hdr = {"Authorization": f"Bearer {token}"} if token else {}
```
nice catch! I see this documented [here](https://docs.github.com/en/actions/security-guides/automatic-token-authentication). However, I might change the approach to use `os.environ.get("GITHUB_TOKEN")` internally rather than taking a `--token` argument. I can handle this later, however, so keep this approach for now.

w-bonelli replied:
Ok, sounds good- that does sound better. I guess end users aren't likely to hit rate limits, only CI.

w-bonelli replied:
If your PR is merged before this one I'll rebase to get the changes you've made in #1473 and use that approach in the CI configs.

w-bonelli commented:
I think this is about ready for review. Updates introduced in this PR are detailed below. Any feedback and criticism is very welcome especially as the scope of changes here grew a bit larger than I originally anticipated.

## Pytest fixtures

This PR builds on [1450](https://github.com/modflowpy/flopy/pull/1450) to make fuller use of `pytest` capabilities. These mainly include keepable temporary directories and shared fixtures for locating and retrieving example data. (Some of the below is reproduced from 1450.)

### Keepable temporary directories

`pytest` provides the [built-in `tmp_path` fixture](https://docs.pytest.org/en/latest/how-to/tmp_path.html#the-tmp-path-fixture) for function-scoped temporary directories, located in the [system's default base temporary directory](https://docs.pytest.org/en/latest/how-to/tmp_path.html#the-default-base-temporary-directory). By default, outputs from the last 3 test sessions are stored in this location, and `pytest` will automatically remove artifacts created by older sessions. This is not ideal for manually inspecting test outputs as the system temporary directory may be inconveniently located or hidden and it's easy to accidentally lose test data due to automatic cleanup.

Previously the `--keep` command line argument addressed this problem. This PR adds a `--keep <dir>` CLI argument for `pytest` with equivalent functionality (the only different is that a path must be provided explicitly). When provided, test outputs will be saved from automatic cleanup at the given location.

Fixtures for function-, class-, module-, and session-scoped temporary directories are provided in `autotest/conftest.py`, all of which are compatible with the `--keep` argument. The fixtures are named according to scope:

- `tmpdir`: function-scope (overrides `pytest`'s deprecated [`tmpdir`](https://docs.pytest.org/en/latest/reference/reference.html#std-fixture-tmpdir))
- `class_tmpdir`: class-scope
- `module_tmpdir`: module-scope
- `session_tmpdir`: session-scope

The `--keep` argument can be shortened to `-K` when invoking `pytest`. The argument's value (e.g. `some/safe/path`) is the path to move test outputs to. If `--keep` is provided, the contents of temporary directories will be copied to subdirectories of `some/safe/path` (each one named according to the corresponding test function, class, module or session) before temporary directories are cleaned up. If `some/safe/path` doesn't exist, it and any necessary parents will be created. If a subdirectory of `some/safe/path` already contains files with the same name as incoming test outputs, the existing files will not be overwritten and an error will be thrown. (Filename conflicts are unlikely since subdirectories of `some/safe/path` will be created/named according to the tests they correspond to, but still probably best to be safe?)

## File reorganization

Test files were previously named in increasing order with pattern `t***_test.py`. Files are now named according to purpose, with some previously colocated tests moved to suitably named files and some related but previously scattered tests now grouped together. (Please let me know if any names or groupings should be reconsidered.)

I also updated `pytest.ini` to match both `test_*.py` and `*_test.py` (previously only the latter matched).

The functions of the previous CI framework can be matched with `pytest` and so the following are removed:

- `ci_cleanup.py`
- `ci_framework.py`
- `ci_prepare.py`

A few outdated files are also removed:

- `readme.txt` was empty
- `runtests.bat` looks to be from `nosetest` days

Example scripts, tutorials and notebooks are now included in the standard test suite, with the corresponding scripts renamed:

- `run_scripts.py` -> `test_example_scripts.py`
- `run_notebooks.py` -> `test_example_notebooks.py`

Scripts and notebooks can still be selected and run separately with the `example` marker:

```shell
pytest -v -m "example"
```

Since `jupyter` and `jupytext` are now required to run the full test suite, I included them in the `test` options group in `setup.cfg` and in `etc/environment.yml`.

## Test atomicity

Most tests previously used `FlopyTestSetup` defined in `ci_framework.py` to create and clean up unique temporary workspaces. All tests now use functionally equivalent `pytest` fixtures defined in `conftest.py` and described above. Those few tests which did not previously clean up their outputs now do so.

## Test coverage

In the process of replacing ad hoc path manipulations with shared fixtures for data access, I parametrized some of the tests a bit more liberally. I also included the `flopy_swi2_ex2.py` and `flopy_swi2_ex5.py` example scripts, which were previously excluded for taking too long (I marked the example script tests as slow to make them optional).

The total number of test cases in the suite has increased from ~760 to ~980 (some of these are the example scripts and notebooks having been folded in). I'm not sure exactly how coverage is affected as `pytest-cov` seems to be misbehaving and I haven't had a chance to debug it yet.

Scripts, tutorials and notebooks are now tested in CI on Windows and OSX as well (previously only on Ubuntu).

## Flaky tests

A few tests depend on external network-bound resources, e.g. [spatialreference.org](spatialreference.org) in `EPSG` (one of the notebooks too). These now trap transient errors and don't fail if the resource is unreachable.

A few other tests occasionally or consistently fail in ways I haven't figured out yet. I marked occasionally flaky but mostly reliable ones with the `flaky` marker, which gives them a configurable number of retries on failure (defaulting to 1). Remaining flaky tests include:

- `test_usg.py::test_freyberg_usg`: occasional `forrtl: error (72): floating overflow`
- `test_mt3d.py::test_mfnwt_CrnkNic`: same as above
- `test_plot.py::test_model_dot_plot`: this and other tests using tk sometimes encounter `_tkinter.TclError: Can't find a usable tk.tcl` (or `init.tcl`)... somehow tk/tcl installation gets into a bad state?
- `test_grid.py::test_voronoi_grid`: one of the grids occasionally has an `ncpl` off by 3 (535 instead of 538)

Tests marked with `xfail` include:

- `test_plot.py::test_cross_section_boundary_conditions`: sometimes `plot_bc()` adds a `LineCollection` instead of a `PatchCollection` to `ax`
- `regression/test_mf6.py::test_np001`: consistent `list index out of range` on Python 3.7, seems to be in `flopy.utils.datautil.PyListUtil`

I also xfailed two tests relating to the way the `Modflow` object handles `external_path`.

- `test_modflow.py::test_mflist_external`
- `test_modflow.py::test_single_mflist_entry_load`

These 2 passed as previously configured, but only by using `"."` or a simple string as relative path. Any absolute paths crash. It looks like the issue may be in [the way relative paths are recomputed in `_reset_external`](https://github.com/modflowpy/flopy/blob/eb59e104bc403387c62649f47816e5a7896d792f/flopy/mbase.py#L1253) when the model workspace changes. I wasn't sure if this was deliberate and we can/should only support relative paths, so I marked it to be safe.

### Eliminating shared state in example scripts & notebooks

Many of the example scripts, tutorials and notebooks previously used subdirectories of the current working directory named either `data` or `temp` as a workspace. This produced intermittent failures if scripts and notebooks were run in parallel, as each notebook assumed sole ownership and wrote to and/or cleaned the location. Scripts, tutorials and notebooks now use unique temporary directories and make a best-effort attempt to clean up afterwards (though this sometimes fails on Windows due to [limitations of Python's `TemporaryDirectory`](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory)). Scripts with a command line interface now have an extra `--keep <dir>` option to match the `pytest` option described above.

## Benchmarking

Previously performance tests were defined in `t064_test_performance.py`. A single model was used with 3 test cases (initialization, load, and write durations) and a hard-coded threshold which was intermittently exceeded on Github CI. This PR introduces `pytest-benchmark` for performance testing. The `benchmark` fixture is injected into any test case requiring it and transparently collects performance statistics about the function it is used to wrap (more [here](https://pytest-benchmark.readthedocs.io/en/latest/)). The original model and test cases were preserved and are now located in `test_modflow.py`:

- `test_simple_model_init_time`
- `test_simple_model_load_time`
- `test_simple_model_write_time`

A number of other tests are now benchmarked as well: all regression tests (anything under `autotest/regression` and using `pymake` to compare outputs) as well as model loading tests exercising various packages.

Benchmarked tests can be identified by the `benchmark` fixture appearing in their signatures. Benchmarking is enabled by default and will appear in stdout, but is not saved to disk. To save JSON output, use `--benchmark-autosave`, which will create a `.benchmark` folder in the current directory. To disable benchmarking, use `--benchmark-disable`. Benchmarking can also be disabled from `pytest.ini` by customizing the `pytest` section:

```toml
[pytest]
addopts = --benchmark-disable
```

Currently benchmarking is disabled on CI. If there is consensus around the `pytest-benchmark` approach I can enable CI benchmarking and save output to a GitHub artifact. The `pytest-benchmark` plugin also has a built-in ability to compare performance results with previously saved statistics, which I have not yet explored. Another option that could be explored is moving benchmarks to a separate CI workflow, so commit-triggered CI finishes faster. Maybe also worthwhile to consider separating the lengthiest tests, which tend to be mf6 example models and a few of the notebooks. For instance, tests for each platform could be split into separate workflows:

- `fast`
- `slow`
- `benchmark`

Where `slow` and `benchmark` could be triggered only when fast tests pass (ideally in a couple minutes max). Or it might make sense to only run benchmarks daily or on some other interval. This might allow a faster feedback cycle and easier development and CI debugging.

Here's a shortlist of the slowest offenders:

- `regression/test_mf6_examples.py::test_mf6_example_simulations`: `ex-gwf-sagehen` and `ex-gwt-rotate` in particular
- `test_example_notebooks.py`: `flopy3_modpath7_create_simulation` and `flopy3_raster_intersection` in particular
- `test_mt3d.py::test_mfnwt_keat_uzf`

## Markers

Markers are provided to mark `slow` tests, `example` tests (which run example scripts, tutorials and notebooks), and `regression` tests. This allows selecting subsets of tests to run when invoking `pytest`.

Markers are also provided to conditionally skip tests if executables or optional dependencies are not available, or if the test is incompatible with a certain operating system. I updated the [developer docs](https://github.com/modflowpy/flopy/blob/9dcf87de6b7795fd8c67bc41325a640a59073806/DEVELOPER.md) with more detailed information about these.

One convenient use for markers is to deselect slow tests. This allows for a faster feedback cycle and (hopefully) encourages frequent testing.

For instance, to run tests in parallel, skipping `slow` tests, with benchmarking disabled (benchmarking is slower as tests are repeated multiple times):

```shell
pytest -v -n auto -m "not slow" --benchmark-disable
```

Running fast tests in this way should complete in under a minute on a reasonably fast machine. In my development environment it takes ~30s for ~650 tests. By comparison, running regression tests (e.g. `pytest -v -n auto -m "regression" --benchmark-disable`) takes ~12m for 141 test cases. The example scripts, tutorials and notebooks (73 test cases) take about 15 minutes, and the full test suite usually about a half hour. (On CI the full suite seems to average slightly over an hour.)

Note that most `slow` tests are either `regression` or `example` tests, but not all, and not all `regression` tests are slow.

## Regression testing and coupled models

Regression tests (tentatively deemed anything using `pymake`'s comparison methods) have been moved to their own `regression` module under `autotest`. This isolates the `mfpymake` dependency and also allows a separate `conftest.py` specifically for regression testing.

Tests using example models downloaded from the [modflow examples repo](https://github.com/MODFLOW-USGS/modflow6-examples) used to be located in `t503_test.py`. They now live in `regression/test_mf6_examples.py::test_mf6_example_simulations`. Previously these tests were implicitly parametrized by simulation directory. Model coupling was achieved by alphabetical ordering of subdirectories in the top-level simulation folder, e.g. `mf6gwf` preceding `mf6gwt`. This assumption is now explicit, with simulations parametrized by ordered list of namfile paths representing a sequence of models to run. Each set of 1+ name files shares the same top-level workspace in order to allow chained models to consume the outputs of prior ones. An arbitrary number of models can be chained in this way.

The parametrization is still generated dynamically from the example model data, as before, but the generation logic has been moved to the `pytest_generate_tests` hook in the `regression` module's `conftest.py`, allowing reuse by other tests. A [file mutual-exclusion lock](https://pypi.org/project/pytest-xdist/#making-session-scoped-fixtures-execute-only-once) is used to synchronize `pytest-xdist` workers and make sure the example data is downloaded only once. Each test case copies relevant subdirectories to an isolated temporary workspace and doesn't alter the cached download location.
   
## Possible next steps

Separating performance/benchmarking from correctness concerns in CI with separate workflows might be a worthwhile next step in the near term?

The model coupling mechanism used for mf6 example models is a rough draft: something could probably be abstracted and reused for similar test cases in Flopy and/or Modflow. The changes here are intended to be broadly compatible with any capabilities provided by modflow devtools in the future (e.g., model-loading fixtures could be updated to point to a git submodule or some other data source). Sharing fixtures between tests should also make it easier to swap in alternative implementations for provisioning or parametrizing tests.

langevin-usgs reviewed (approved):
Nicely done.
